Add initial implementation

morrisallison opened the pull request:
Parents: #1 #2

Additional properties and methods were added, but I only consider `clone()` to be supplementary. I'm working on writing integration tests for a tiny DRF service.

:bell: ping @yola/frontend-engineering

Commit:
Add initial implementation

morrisallison commented:
One point of concern is when calling `next()`, `prev()`, `first()`, `last()`, or `current()` repeatedly without waiting for requests to finish. I was going to add a request queue, but I figured I grab some opinions first.

Commit:
Refactor module imports

Commit:
Refactor the paginator to cache promises instead of responses

Commit:
Update dependency list

LionOps commented on .jshintignore:
```diff
@@ -0,0 +1,33 @@
+# Logs
```
Is there a point to this file? Snitch doesn't need it, and all of our other libs don't have one either. If Snitch did need it, it'd make sense, but I think it can go. If you do see a use case, share.

beck replied:
This file is the same as the gitignore, why not symlink it?

morrisallison replied:
My bad, I was using it for local changes not in this PR yet. It isn't needed (yet?)

LionOps commented on README.md:
```diff
@@ -0,0 +1,155 @@
+import querystring from 'querystring';
+
+
+const request = function(options, queryParams) {
```
I think the example should use `options` to illustrate what we mean.

LionOps commented on README.md:
```diff
@@ -0,0 +1,155 @@
+## Options
+
+* `limitQueryParam`
+  - Matches the endpoint's [`limit_query_param`](http://www.django-rest-framework.org/api-guide/pagination/#configuration_1) option
```
Since this README is going to stick around, let's move all the links into references and have the links defined at the bottom of the page. It'll increase the readability of the markdown and is cleaner in general.

beck replied:
Big fan of using markdown link IDs. They can be defined anywhere, please don't put them at the bottom of the page (is really annoying when editing to scroll up & down). Preference is to keep at the bottom of the section:

```
### Header

Content with [some link][].

[some link]: http://some-link.tld "optional title if needed"

```

LionOps commented on README.md:
````diff
@@ -0,0 +1,155 @@
+  });
+```
+
+### Using two paginators to iterate over ten pages at a time:
````
I found this to be slightly confusing. I assume that each call to `next` returns 10 pages of results, no?

LionOps commented on README.md:
```diff
@@ -0,0 +1,155 @@
+
+## Examples
+
+### Creating a paginator and fetching the first page:
```
My personal preference would be to structure the README in a way that shows example code using all the functions on the `paginator` and then documentation on what each does. You don't have to adopt that style, but no where is `next`, `prev`, or `last` mentioned in the README.

Due to your example use of `first`, is that how you foresaw the paginator primarily to be used? For me, `first` is useless unless I've called `next` a few times in most cases.

LionOps commented on src/helpers/request-spy.js:
```diff
@@ -0,0 +1,51 @@
+import sinon from 'sinon';
+
+
+const limitOffsetResponse = require('../fixtures/limit-offset-response.json');
```
Why not use the import syntax?

beck replied:
The current loader spec only supports importing javascript.

Weather or not this will work:

```
import limitOffsetResponse from '../fixtures/limit-offset-response.json';
```

Will depend on the toolchain:
- :+1: babel > ES2015 > browserify
- :x: SystemJS will fail unless you use a json plugin

Here's a ticket asking if the module loader will ever give JSON special treatment:
https://github.com/whatwg/loader/issues/133

With that said, I'm in favor of using `import`, load the json as a "data module", and keep it at the top alongside all the other module imports.

LionOps replied:
> With that said, I'm in favor of using import, load the json as a "data module", and keep it at the top alongside all the other module imports.

If it depends on toolchain, best to to use `import`. `require` immediately shoehorns consumers into browserify.

morrisallison replied:
Got it, I'll use import going forward.

LionOps commented on src/page-merger.js:
```diff
@@ -0,0 +1,62 @@
+
+  merge(startPage, endPage) {
+    return this._fetchRequests(startPage, endPage)
+      .then(this._makeCollection.bind(this))
```
Consider `reqs => this._makeCollection(reqs)` instead.

Hkly replied:
Why?

LionOps replied:
Double arrow binds context. They're the same, but one doesn't have `bind`.

morrisallison replied:
This is just me sticking to the no anon function rule too strictly. I'm glad we're changing that.

LionOps commented on src/page-merger.js:
```diff
@@ -0,0 +1,62 @@
+  _makeCollection(responses) {
+    const collection = [];
+
+    for (let i = 0; i < responses.length; i++) {
```
This is a good opportunity for `reduce`, either native or lodash's implementation.

Hkly replied:
Would `map` make more sense?

LionOps replied:
Yeah -_-

LionOps commented on src/page-merger.js:
```diff
@@ -0,0 +1,62 @@
+    let count = null;
+    let results = [];
+
+    for (let i = 0; i < collection.length; i++) {
```
`reduce` here as well

LionOps commented on src/paginator-spec.js:
```diff
@@ -0,0 +1,486 @@
+        const requestCallCount = paginator.first()
+          .then(request.getCallCount);
+
+        return expect(requestCallCount).to.eventually.equal(1);
```
Shouldn't this be 2? Once for the given page and once for the first page?

LionOps commented on src/paginator-spec.js:
```diff
@@ -0,0 +1,486 @@
+      });
+
+      it('calls the request given to the paginator', function() {
+        const requestCallCount = paginator.last()
```
Calling `.last` instead of `.current`

LionOps commented on src/paginator-spec.js:
```diff
@@ -0,0 +1,486 @@
+
+        const currentPage = paginator.next().then(getCurrentPage);
+
+        return expect(currentPage).to.eventually.equal(2);
```
This is wrong. The first call to `next` should return a promise to the current page. Subsequent calls should increment it.

We are purposefully implementing this similar to an iterator, in which the first call to `next` returns the first value. `first` should be considered a convenience method and not the primary way to get the first page. 

The first element shouldn't require any special treatment. Consider:

``` js
let firstFivePagePromises = [];

firstFivePagePromises.push(paginator.first());

for (let i= 0; i < 4; i++) {
  firstFivePagePromises.push(paginator.next());
}
```

vs

``` js
let firstFivePagePromises = [];

for (let i= 0; i < 5; i++) {
  firstFivePagePromises.push(paginator.next());
}
```

For `first`, it it's to reset the paginator to the beginning, it should set it to page `0`;

LionOps replied:
Likewise, the constructor should decrement the page given or set it to 0.

LionOps replied:
Or you're incrementing after fetching the page, in which case current page makes sense.... Can we get a test that asserts that the response is for the first page?

morrisallison replied:
Okay got it, I misinterpreted the design.

LionOps commented on src/paginator-spec.js:
```diff
@@ -0,0 +1,486 @@
+        return expect(requestCallCount).to.eventually.equal(1);
+      });
+
+      it('clone responses', function() {
```
Hmm. This seems unintuitive. Every other method shouldn't be making duplicate requests, instead caching them and returning that for duplicate calls.

For instance, if I'm on page 2 and call `next`, `next`, `prev`, `prev`, `next` I'd expect 2 requests to be made and both `prev` calls and the last `next` call to return caches. I think `fetchPage` should be the same.

LionOps commented on src/paginator-spec.js:
```diff
@@ -0,0 +1,486 @@
+        return getResponses.spread(assert.notStrictEqual);
+      });
+
+      it('caches responses', function() {
```
Wait... Maybe I misunderstood. Are you caching the request promise and then cloning what `fetchPage` returns so that consumers never affect the cached promise?

morrisallison replied:
Correct, it's caching the Promises, and cloning the responses.

LionOps commented on src/paginator.js:
```diff
@@ -0,0 +1,300 @@
+      }
+    };
+
+    return this.fetchPageCount()
```
I think this is a race condition. You should be incrementing the page before making the request, else two calls to `next` will return the same page.

I think you'll need to decrement the `page` during instantiation and then this will increment up to the desired page. As a result, `first` should set the page to `0`

LionOps commented on src/paginator.js:
```diff
@@ -0,0 +1,300 @@
+    const queryParams = this._makeQueryParams(page);
+    const opts = this._options;
+
+    return new Paginator(req, reqOpts, queryParams, opts);
```
I'm not sure it matters, but it does seem weird to clone the paginator without cloning the things passed into it. I would clone these too so there is a true clone that doesn't share internal references to the same objects.

morrisallison replied:
Agreed, it doesn't affect anything, but passing in the same values isn't good.

LionOps commented:
> One point of concern is when calling next(), prev(), first(), last(), or current() repeatedly without waiting for requests to finish. I was going to add a request queue, but I figured I grab some opinions first.

How would the request queue work? I think that calling them repeatedly should return a promise for the next page. There is a risk that next is called more times than actual pages, but I think we can have documented behavior around that. Calling next() on a page that doesn't exist should resolve to the same error regardless if it's a known limit or not. Knowing the page count should merely prevent the request actually being made.

beck commented on README.md:
```diff
@@ -0,0 +1,155 @@
+
+
+const limit = 50;
+const style = Paginator.styles.limitOffset;
```
Why does the paginator need a configured "style"?  Shouldn't the behavior be determined by given options?

morrisallison replied:
The `style` option determines what `queryParams` are sent to the request. Requests aren't supposed to interpret the query params, but only serialize them. This way the paginator is configured to match the request rather than the other way around.

beck commented on README.md:
```diff
@@ -0,0 +1,155 @@
+
+const limit = 50;
+const style = Paginator.styles.limitOffset;
+const paginator = new Paginator(request, null, {limit}, {style});
```
I need a general overview of the API before getting here, otherwise I have no clue what `null` means.

beck commented on src/paginator.js:
```diff
@@ -0,0 +1,300 @@
+
+
+export const styles = Object.freeze({
+  pageNumber: 1,
```
This looks like an enum, I generally expect the members to be uppercase

beck replied:
also, `pageNumber` is a misleading label - especially when the value is an actual number

Hkly replied:
Curious, why is it misleading that `pageNumber` contains a number?

beck replied:
Because the value of `pageNumber` is not at all a number of a page.  It is an int representing a style of pagination.

I generally find enums to be a weak code smell—they're often used for swtich statements.  I view switch statements as a code smell because it hints that we're not using good OO. For example: instead of creating a single class with behavior that branches off an enum: create two classes: `PageNumberPaginator` and `LimitOffsetPaginator` (and you can see that is [how it is done in drf](http://www.django-rest-framework.org/api-guide/pagination/#pagenumberpagination)). Now we have eliminated branching logic in our source (a big win) and the consumer will decide which class to use.

I'm not saying this is what must be done. To iron out the complexity there may be a functional, less OO, approach that is more javascript friendly. Always more than one way to skin a cat.

Hkly replied:
> Because the value of pageNumber is not at all a number of a page. It is an int representing a style of pagination.

Oo I see. It's a style that just happens to be called "pageNumber". Got it, thanks!

Hkly replied:
It would maybe improve readability if these were both called `pageNumberPagination` and `limitOffsetPagination` or something along those lines to clearly indicate what they are.

LionOps replied:
> I generally find enums to be a weak code smell—they're often used for swtich statements. I view switch statements as a code smell because it hints that we're not using good OO. For example: instead of creating a single class with behavior that branches off an enum: create two classes: PageNumberPaginator and LimitOffsetPaginator (and you can see that is how it is done in drf). Now we have eliminated branching logic in our source (a big win) and the consumer will decide which class to use.

I semi-disagree with switch statements being a code-smell, but....

Following SOLID, the S (Single Responsibility) is violated by not having two different classes based on the style. We should definitely break this up into a parent and child classes.

(And doing so means `Paginator` takes only three arguments)

beck replied:
Ohh if you're a fan of SOLID, let me tell a story about why switch statements are harmful to software structure.

One upon a time...

drf-paginator is published and the software looks like:

```
Paginator
  Switch
      Style A
      Style B
```

Hannah has need for a paginator in her personal project. She uses a custom flavor of DRF and needs to use `Style H`. She starts writing the new pagination style and is soon disappointed, she needed to edit our source code to make her customization.

The design is bad—where did it go wrong?

`Paginator` calls `Style A`, the dependency points from `Paginator > Style A`. A new style is not able to be written without forking and editing the lib, the switch statement has destroyed independent deployability.

Hannah, who hates switch statements, refactors.  Opens a pull for:

```
Paginator

Style A
  Paginator

Style B
  Paginator
```

The dependencies now point up, `Style A > Paginator` (DI). Paginator is now open for extension without needing to edit (Open/closed). In her project, she's able to include `drf-paginator.js` as a dependency and write `style-h.js` separately. Updating `style-h.js` does not require bumping `drf-paginator.js`. Independent deployability has been restored.

The pull is gladly accepted since it is a much better design.

Final note: switch statements are not always bad, they're often used when writing MAIN. All dependencies continue to point toward the application. So long as the dependencies point in the right direction, switch statements are safe.

morrisallison replied:
> This looks like an enum, I generally expect the members to be uppercase

Yup, it is. I agree on uppercase too, though I haven't seen it being used. That's one the things I listed in the listed in the design guide update. I'll mention it there.

beck commented on src/paginator-spec.js:
```diff
@@ -0,0 +1,486 @@
+      });
+    });
+
+    describe('clone', function() {
```
Is `clone` a use case we expect to have? Questioning it because it seems like a feature not required for an initial implementation

morrisallison replied:
It's not needed. We decided that the paginator wouldn't have a method to reset cache, but instead we'd just create a new paginator instance. This was a quick helper to facilitate that.

beck commented on src/paginator.js:
```diff
@@ -0,0 +1,300 @@
+};
+
+export class Paginator {
+  constructor(request, requestOptions = null, queryParams = {}, options = {}) {
```
This function signature is rough.

beck replied:
Would it not work with just:
- request
- options: page, limitOffset

LionOps replied:
We wanted to make it match up with a request's parameters. A request takes two, `options` and `queryParams`, so Paginator should take at least three.

I realize this has four, but see my other comment I'm about to make.

beck commented:
I'm a little surprised a class is used instead of a more functional approach. Notice that there is a proposal for [Async Iterator](https://github.com/tc39/proposal-async-iteration) for which FB has [already written a polyfill](https://github.com/facebook/regenerator/blob/f87d654f85c9925c4db3f74806f7615a71297f40/runtime.js#L136)? Not saying this is the needed approach, but it would feel more contemporary.

beck commented on README.md:
```diff
@@ -0,0 +1,155 @@
+  - Indicates which pagination style to use.
+  - Currently supports:
+    - [`PageNumberPagination`](http://www.django-rest-framework.org/api-guide/pagination/#pagenumberpagination)
+    - [`LimitOffsetPagination`](http://www.django-rest-framework.org/api-guide/pagination/#limitoffsetpagination)
```
Testing our endpoints, do we support `LimitOffsetPagination` anywhere? Supporting alternate pagination styles seems like a feature that could come after the initial release.

morrisallison replied:
No, it's not needed now. However, when writing the paginator, I wanted to ensure it was flexible enough to adapt to the other styles. Taking it out now wouldn't change much of the code. I also saw this https://github.com/yola/production/issues/3061#issuecomment-193303279

Commit:
Merge master

Commit:
Refactor paginator into seperate modules

snitch commented:
There are 22 errors:

* src/paginators/page-number-paginator-spec.js, line 6 - Line is too long.
* src/paginators/limit-offset-paginator-spec.js, line 38 - Line is too long.
* src/paginators/base-paginator.js, line 102 - Line is too long.
* src/paginators/limit-offset-paginator-spec.js, line 10 - 'assert' is defined but never used.
* src/paginators/limit-offset-paginator-spec.js, line 6 - Line is too long.
* src/query-handlers/page-number-query-handler-spec.js, line 7 - 'assert' is defined but never used.
* src/paginators/base-paginator.js, line 1 - 'assign' is defined but never used.
* src/query-handlers/limit-offset-query-handler-spec.js, line 37 - Line is too long.
* src/paginators/limit-offset-paginator.js, line 2 - Line is too long.
* src/paginators/page-number-paginator-spec.js, line 35 - Line is too long.
* src/query-handlers/limit-offset-query-handler-spec.js, line 94 - Line is too long.
* src/query-handlers/limit-offset-query-handler-spec.js, line 7 - 'assert' is defined but never used.
* src/paginators/base-paginator-spec.js, line 7 - 'PaginatorError' is defined but never used.
* src/query-handlers/page-number-query-handler-spec.js, line 4 - Line is too long.
* src/query-handlers/index.js, line 1 - Line is too long.
* src/paginators/base-paginator-spec.js, line 3 - 'sinon' is defined but never used.
* src/paginators/page-number-paginator-spec.js, line 10 - 'assert' is defined but never used.
* src/paginators/base-paginator.js, line 3 - 'omit' is defined but never used.
* src/paginators/base-paginator-spec.js, line 326 - Line is too long.
* src/query-handlers/limit-offset-query-handler-spec.js, line 4 - Line is too long.
* src/query-handlers/index.js, line 2 - Line is too long.
* src/query-handlers/page-number-query-handler-spec.js, line 52 - Line is too long.

snitch commented:
:sparkles: No lint errors found. :sparkles:

morrisallison commented:
- README and doc updates are next
- The paginator has been split into several modules.
- The `options` argument in the constructor has been removed.
  - Query parameter handling is now configured separately.
- The package now throws `PaginatorError` instead of `Error`.
  - `PaginatorError` extends `Error` the old fashion way
  - [`babel-plugin-transform-builtin-extend`](https://www.npmjs.com/package/babel-plugin-transform-builtin-extend) didn't work properly

Commit:
Minor fixes

Remove leftover unused property.
Correct `LimitOffsetQueryHandler` class name.
Add newline to end of limit-offset-query-handler-spec.
Regress to exporting paginators and query handlers from the root module.

Commit:
Update npm scripts

Add script to run jshint.
Add script to run unit tests.
Update the test script to run jshint.

Commit:
Update documentation

Update README
Add API reference

snitch commented:
:sparkles: No lint errors found. :sparkles:

morrisallison commented:
Links to docs
- [README](https://github.com/yola/drf-paginator/blob/morris/initial/README.md)
- [API](https://github.com/yola/drf-paginator/blob/morris/initial/API.md)

Commit:
Remove unused files

LionOps commented on src/page-merger-spec.js:
```diff
@@ -0,0 +1,62 @@
+      });
+
+      it('returns a promise that successfully resolves', function() {
+        const promsie = pageMerger.merge(1, pageCount);
```
Promsie? I'm imagining Jar Jar Binks now asking Padma to the prom now. "Yousa go to the promsie withsa me?"

LionOps commented on src/paginators/base-paginator-spec.js:
```diff
@@ -0,0 +1,361 @@
+        return expect(promise).to.eventually.resolve;
+      });
+
+      it('returns a promise with the a response object', function() {
```
with the a

Commit:
Add coverage reporting

LionOps commented on src/paginators/base-paginator-spec.js:
```diff
@@ -0,0 +1,361 @@
+        const requestCallCount = paginator.last()
+          .then(request.getCallCount);
+
+        return expect(requestCallCount).to.eventually.equal(2);
```
Why is this 2 as opposed to 1? I'm guessing it's because paginator requests the first page (or whatever is set in the options) before making a request for the last page. Is this correct?

LionOps commented on src/paginators/base-paginator-spec.js:
```diff
@@ -0,0 +1,361 @@
+
+      it('returns a rejected promise for pages after the last page',
+        function() {
+          const promise = paginator.fetchPage(-3);
```
Wouldn't -3 be before the first page as opposed to after the last page?

LionOps commented on src/paginators/base-paginator.js:
```diff
@@ -0,0 +1,133 @@
+    const handleResponse = (response) => {
+      return this._handleResponse(response, page);
+    };
+    const request = this._sendRequest(page)
```
linebreak

LionOps commented on src/paginators/base-paginator.js:
```diff
@@ -0,0 +1,133 @@
+      return this._getRequest(page)
+        .then(cloneDeep);
+    }
+    if (!this._isValidPage(page)) {
```
linebreak

LionOps commented on src/paginators/base-paginator.js:
```diff
@@ -0,0 +1,133 @@
+    this.queryHandler = null;
+  }
+
+  get page() {
```
Is this a typo or an es6 thing?

morrisallison replied:
ES6, exposing `_page` while not making it writable.

euwest replied:
> ES6, exposing _page while not making it writable.

"it" in that sentence is referring to `_page` or `page`? I'm guessing `page`, since `_page` has an implicit getter and setter, the way you built it in the constructor

LionOps commented on src/query-handlers/limit-offset-query-handler-spec.js:
```diff
@@ -0,0 +1,123 @@
+          expect(queryParams).to.be.an('object');
+        });
+
+        it('throws an error when a limit hasen\'t been set', function() {
```
Spelled "hasn't" wrong

beck replied:
```
hasn’t
```

LionOps commented on src/query-handlers/limit-offset-query-handler.js:
```diff
@@ -0,0 +1,122 @@
+    if (offset === 0) {
+      return 1;
+    }
+    if (!limit) {
```
linebreak

LionOps commented:
The design looks solid, I think.

Commit:
Tweak docs

LionOps commented on API.md:
````diff
@@ -0,0 +1,445 @@
+
+```javascript
+let paginator = new BasePaginator(request);
+// True
````
Slight nitpick, but can the comment come after the statement? My preference is based solely on the fact that it mimics what one would see as part of a REPL.

morrisallison replied:
Sure, it makes more sense that way since it's a result

stefanor replied:
You could say `-> True` or `returns: True` to make the meaning of the comment more obvious

LionOps commented on API.md:
```diff
@@ -0,0 +1,445 @@
+let request = function(options, queryParams) {
+  let {user} = options;
+  let query = querystring.stringify(queryParams);
+  let url = `https://example.com/users/${user/}?${query}`;
```
Is the / in `${user/}` correctly placed?

Commit:
Update package.json

Move Babel packages to devDependencies
Remove directories property

Commit:
Remove unused whatwg-fetch devDependency

Commit:
Correct invalid page handling

Commit:
Correct test descriptions and typos

snitch commented:
:sparkles: No lint errors found. :sparkles:

Commit:
Update comments in API doc to be REPL like

LionOps commented on package.json:
```diff
@@ -30,31 +28,32 @@
   },
   "devDependencies": {
+    "babel-preset-es2015": "~6.6.0",
+    "babelify": "~7.2.0",
```
These should be in deps.

Commit:
Remove unused test code

Commit:
Rename variables for clarity

Commit:
Add default values to the query handlers

Commit:
Simplify resolvePage to a single purpose

Commit:
Add tests for 100% coverage

Commit:
Remove Karma

Remove Karma dependencies
Update test scripts
Move babel-preset-es2015 to dependencies
Add app-module-path dependency

snitch commented:
:sparkles: No lint errors found. :sparkles:

morrisallison commented:
The reported code coverage wasn't correct, now it is. For better or worse, the tests themselves are also included in the coverage results.

Commit:
Add line breaks

snitch commented:
:sparkles: No lint errors found. :sparkles:

LionOps commented on API.md:
```diff
@@ -50,14 +50,13 @@
-This is a [getter][mdn-defineproperty] property.
-
-[mdn-defineproperty]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/defineProperty
+This is a property is read-only.
```
Grammarize

LionOps commented on API.md:
```diff
@@ -0,0 +1,437 @@
+
+#### paginator.page
+
+This is a property is read-only.
```
Github hid it. Grammarize this.

Commit:
Correct grammar

Commit:
Update invalid page handling to reject an error

snitch commented on src/paginators/base-paginator.js:
```diff
@@ -0,0 +1,143 @@
+  _rejectPage(page) {
+    const message = `Invalid page "${page}".`;
+    const error = new Error(message);
+    error.response = { detail: message }
```
Missing semicolon.

LionOps replied:
I think this also caused tests to break.

LionOps commented on package.json:
```diff
@@ -5,8 +5,14 @@
+    "cover": "rm -rf coverage && babel-node node_modules/.bin/isparta cover --root src/ --report text --report lcov --report html node_modules/.bin/_mocha -- src/test-setup.js $(find src -name '*-spec.js')",
+    "coveralls": "cat coverage/lcov.info | coveralls",
+    "jshint": "jshint src",
+    "test": "npm run cover && npm run jshint",
```
I don't think we want linting to break tests. Snitch does that for us, and because linting errors are often stylelistic, they shouldn't affect tests.

Do you agree?

morrisallison replied:
I remember @thomasw saying we ought to, but didn't find a comment after some searching. It is ran in SBUI though.

Commit:
Add missing semicolon

snitch commented:
:sparkles: No lint errors found. :sparkles:

LionOps commented:
👍  (since @euwest is also reviewing, let's get a second thumbs! It's a huge PR.)

euwest commented on src/actions.js:
```diff
@@ -0,0 +1,32 @@
+  return null;
+};
+
+export const parseResponse = function(response) {
```
I'm curious why this is called `parseResponse` when it seems to just be adding page count and limit attributes to the response?

morrisallison replied:
I called it parse since it's only pulling out the properties of the response it cares about, and creating a new object.

However, I do see where you're coming from. The function is really only calculating the `pageCount`. I'll refactor it to do only that.

euwest commented on src/actions.js:
```diff
@@ -0,0 +1,32 @@
+    count,
+    limit,
+    pageCount,
+    results
```
Could you explain what `count`, `limit`, and `pageCount` are/how they differ?

morrisallison replied:
This will addressed in the refactor [mentioned here](https://github.com/yola/drf-paginator/pull/3#discussion_r60782819).

Commit:
Simplify actions

snitch commented:
:sparkles: No lint errors found. :sparkles:

euwest commented on src/actions.js:
```diff
@@ -0,0 +1,26 @@
+  const count = response.count;
+  const resultCount = response.results.length;
+
+  if (count > resultCount) {
```
what's a scenario where this is true, and why would that arise?

morrisallison replied:
When a query has multiple pages, that expression will always be true for the first page response.
The times when it's not true, is on the last page, or when there are no results.

Commit:
Change absolute modules paths to be relative

Commit:
Remove unneeded app-module-path dependency

Commit:
Remove the browserify paths option

euwest commented:
> For better or worse, the tests themselves are also included in the coverage results.

I'm guessing that's a side effect of having the tests mixed in with the implementation files.

snitch commented:
:sparkles: No lint errors found. :sparkles:

morrisallison commented:
> I'm guessing that's a side effect of having the tests mixed in with the implementation files.

Yeah, I tried a bunch of different ways and just settled on that. I don't mind.

euwest commented:
👍 let's make a ticket for fleshing out the readme and API docs. as is, they aren't helpful if you don't know anything about the module.

LionOps commented:
Closes #1 and #2
